Overhaul ecosystem overview (#75)
esp-rs/book: src/overview/bare-metal.md

# Developing on Bare Metal (`no_std`)

> Using `no_std` may be more familiar to embedded Rust developers; it does not use `std` (the Rust standard library) but instead uses a subset, the `core` library. [The Embedded Rust Book][embedded-rust-book] has a great [section][embedded-rust-book-no-std] on this.

Reviewer: Also, core is using backtick, but standard is not, since they are both libraries names, we should use the same format

> Using `no_std` may be more familiar to embedded Rust developers; it does not use `std` (the Rust standard library) but instead uses a subset, the `core` library. [The Embedded Rust Book][embedded-rust-book] has a great [section][embedded-rust-book-no-std] on this.

Author: Added the links and fixed the formatting for standard.

> It's important to note that in general a `no_std` crate can always compile in `std` environment but the inverse is not true. Therefore, when creating crates it's worth keeping in mind if it needs the standard library to function.

Reviewer: ```suggestion
It's important to note that since `no_std`uses the Rust core library, a subset of the Rust standard library,  a `no_std` crate can compile in `std` environment but the oposite is not true. Therefore, when creating crates it's worth keeping in mind if it needs the standard library to function.

```

[embedded-rust-book]: https://docs.rust-embedded.org/
[embedded-rust-book-no-std]: https://docs.rust-embedded.org/book/intro/no-std.html

> ## Relevant `esp-rs` crates

Reviewer: ```suggestion
## Current support
```

> The table below covers what `no_std` features are supported by Espressif products.

Reviewer: ```suggestion
The table below covers the current support for `no_std` at this momment for the different Espressif products.
```

> **Notes**:
>

> > - Hover your cursor over the links in the table header to see feature details.

Reviewer: ```suggestion
```

I would remove this comment, in my opinion, seeing a link on it should be self-explanatory, and we also have an explanation on the list bellow

Author: Removed

> > - The icon ⏳ indicates that the feature is under development.

Reviewer: If we explain what ⏳ indicates, we probably should explain the others emojis (?)

Author: This is exactly what I wanted to hear! :)

Then I will leave the description as it is now. The features to be developed in the future are, kind of, under long-term development.

> > - Features marked with * must be installed for `no_std` to work.

Reviewer: We do not need to install `esp-hal` and `esp-pacs`.  So I would remove the `*` from the table and this note.

Author: Done

|         | [HAL][esp-rs/esp-hal]* | [PAC][esp-rs/esp-pacs]* | [Wi-Fi][esp-rs/esp-wifi] | [Bluetooth][esp-rs/esp-wifi] | [Heap][esp-rs/esp-alloc] | [Println][esp-rs/esp-println] | [Backtrace][esp-rs/esp-backtrace] | [Storage][esp-rs/esp-storage] |
| -------- | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
|  ESP32   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |
| ESP32-C2 |   ✅   |   ✅   |   ✅   |   ✅   |   ❌   |   ✅   |   ✅   |   ✅   |
| ESP32-C3 |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |
| ESP32-C6 |   ✅   |   ✅   |   ✅   |   ⏳   |   ❌   |   ✅   |   ✅   |   ❌   |
| ESP32-S2 |   ✅   |   ✅   |   ✅   |   ❌   |   ✅   |   ✅   |   ✅   |   ✅   |
| ESP32-S3 |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |
| ESP32-H2 |   ❌   |   ✅   |   ❌   |   ❌   |   ❌   |   ✅   |   ❌   |   ❌   |
| ESP8266  |   ✅   |   ✅   |   ❌   |   ❌   |   ❌   |   ✅   |   ❌   |   ❌   |

> The `no_std` features are elaborated on below.

Reviewer: ```suggestion
Here is a list with the relevant crates and small summary of what they do for `no_std`:
```

- **HAL** -- Hardware abstraction layer, see [esp-rs/esp-hal] and [esp-rs/esp8266-hal]
- **PAC** -- Peripheral access crates, see [esp-rs/esp-pacs]
- **Wi-Fi** -- Wi-Fi support, see [esp-rs/esp-wifi]
- **Bluetooth** -- Bluetooth LE support, see [esp-rs/esp-wifi]
- **Heap** -- Simple heap allocator, see [esp-rs/esp-alloc]
- **Println** -- `print!`, `println!`, see [esp-rs/esp-println]
- **Backtrace** -- Exception and panic handlers, see [esp-rs/esp-backtrace]

> - **Storage**  -- Embedded-storage traits to access unencrypted flash memory, see [esp-rs/esp-storage]

Reviewer: Maybe change the list to be crate centered, eg:
- esp-hal: Hadwared abstraction layers
- esp-wifi: Support for wifi, ble, esp-now...
- ...

Author: @SergioGasquez 

Then how about returning to a table format like this? We also use a similar format for `std` crates/repositories.

Should the first column be called repository or feature or crate? Also relevant for `std` crates/repositories.

| Repository | Symbol | Description |
| ------- | :----: | ----------- |
| [esp-rs/esp-hal] | **HAL** | Hardware abstraction layer |
| [esp-rs/esp-pacs] | **PAC** | Peripheral access crates |
| [esp-rs/esp-wifi] | **Wi-Fi** | Wi-Fi support |
| [esp-rs/esp-wifi] | **Bluetooth** | Bluetooth LE support |
| [esp-rs/esp-alloc] | **Heap** | Simple heap allocator |
| [esp-rs/esp-println] | **Println** | `print!`,  `println!` |
| [esp-rs/esp-backtrace] | **Backtrace** | Exception and panic handlers |
| [esp-rs/esp-storage] | **Storage** | Embedded-storage traits to access unencrypted flash memory |

[esp-rs/esp-hal]: https://github.com/esp-rs/esp-hal "Hardware abstraction layer"
[esp-rs/esp8266-hal]: https://github.com/esp-rs/esp8266-hal
[esp-rs/esp-pacs]: https://github.com/esp-rs/esp-pacs "Peripheral access crates"
[esp-rs/esp-wifi]: https://github.com/esp-rs/esp-wifi "Wi-Fi and Bluetooth LE support"
[esp-rs/esp-alloc]: https://github.com/esp-rs/esp-alloc "Simple heap allocator"
[esp-rs/esp-println]: https://github.com/esp-rs/esp-println "print!, println!"
[esp-rs/esp-backtrace]: https://github.com/esp-rs/esp-backtrace "Exception and panic handlers"
[esp-rs/esp-storage]: https://github.com/esp-rs/esp-storage "Embedded-storage traits to access unencrypted flash memory"

Reviewer: > Then how about returning to a table format like this? 

I think its okay to have such table, but I would remove the Symbol column

> Should the first column be called repository or feature or crate? 

I would leave the "repository" term because one repo can contain several crates.

Author: Decided to keep column **Symbol** and make it first in the table.

If we find a way to make mapping between the table above and this one easier without symbols, we can always delete this column to keep this tables formatting the same as its `std` counterpart.

# Developing on Bare Metal (`no_std`)

Using `no_std` may be more familiar to embedded Rust developers; it does not use `std` (the Rust [`standard`][rust-lib-std] library) but instead uses a subset, the [`core`][rust-lib-core] library. [The Embedded Rust Book][embedded-rust-book] has a great [section][embedded-rust-book-no-std] on this.

> It's important to note that since `no_std`uses the Rust core library, a subset of the Rust standard library,  a `no_std` crate can compile in `std` environment but the oposite is not true. Therefore, when creating crates it's worth keeping in mind if it needs the standard library to function.

Reviewer: ```suggestion
It's important to note that since `no_std`uses the Rust `core` library, a subset of the Rust `standard` library,  a `no_std` crate can compile in `std` environment but the oposite is not true. Therefore, when creating crates it's worth keeping in mind if it needs the `standard` library to function.
```

Reviewer 2: > `no_std`uses

MIssing space there

> Rust `standard` library`

I think it's better to be consistent and use either "Rust standard library" or `std`

Reviewer: > I think it's better to be consistent and use either "Rust standard library" or std

This is something that should probably also be included in the guidelines (https://github.com/esp-rs/book/pull/79)

Reviewer 2: Didn't even know that PR existed, will take a look

> It's important to note that since `no_std`uses the Rust core library, a subset of the Rust standard library,  a `no_std` crate can compile in `std` environment but the oposite is not true. Therefore, when creating crates it's worth keeping in mind if it needs the standard library to function.

Reviewer: ```suggestion
It's important to note that since `no_std`uses the Rust core library, a subset of the Rust standard library,  a `no_std` crate can compile in `std` environment, but the opposite is not true. Therefore, when creating crates, it's worth keeping in mind if it needs the standard library to function.
```

| -------- | :-------------------: | :----------------------------------: | :-------------------------------: | :---------------------------: |
| ESP32    |           ✅           |                  ✅                   |                 ✅                 |               ✅               |
| ESP32-C2 |           ✅           |                  ✅                   |                 ✅                 |               ✅               |
| ESP32-C3 |           ✅           |                  ✅                   |                 ✅                 |               ✅               |
| ESP32-C6 |           ✅           |                  ✅                   |                 ✅                 |               ✅               |
| ESP32-S2 |           ✅           |                  ✅                   |                 ✅                 |               ✅               |
| ESP32-S3 |           ✅           |                  ✅                   |                 ✅                 |               ✅               |
| ESP32-H2 |           ⏳           |                  ⏳                   |                 ⏳                 |               ⏳               |

> **Note**:
>

> > - ✅ in Wi-Fi/BLE/ESP-NOW means that the target supports, at least, one the listed technologies. For details see [Current support][esp-wifi-current-support] table of the esp-wifi repository.

Reviewer: ```suggestion
> - ✅ in Wi-Fi/BLE/ESP-NOW means that the target supports, at least, one of the listed technologies. For details, see [Current support][esp-wifi-current-support] table of the esp-wifi repository.
```

> - [ESP8266 HAL][esp-rs/esp8266-hal] is in maintenance mode and no further development will be done for this chip.

[esp-wifi-current-support]: https://github.com/esp-rs/esp-wifi#current-support
### Relevant `esp-rs` crates

| Repository             | Description                                                |
| ---------------------- | ---------------------------------------------------------- |
| [esp-rs/esp-hal]       | Hardware abstraction layer                                 |
| [esp-rs/esp-pacs]      | Peripheral access crates                                   |
| [esp-rs/esp-wifi]      | Wi-Fi, BLE and ESP-NOW support                             |
| [esp-rs/esp-alloc]     | Simple heap allocator                                      |
| [esp-rs/esp-println]   | `print!`,  `println!`                                      |
| [esp-rs/esp-backtrace] | Exception and panic handlers                               |

> | [esp-rs/esp-storage]   | Embedded-storage traits to access unencrypted flash memory |

Reviewer: ```suggestion
| [esp-rs/esp-storage]   | Embedded-storage traits to access unencrypted flash memory |

### When you might want to use bare metal (`no_std`)

- Small memory footprint: If your embedded system has limited resources and needs to have a small memory footprint, you will likely want to use bare-metal because `std` features add a significant amount of final binary size and compilation time.
- Direct hardware control: If your embedded system requires more direct control over the hardware, such as low-level device drivers or access to specialized hardware features you will likely want to use bare-metal because `std` adds abstractions that can make it harder to interact directly with the hardware.
- Real-time constraints or time-critical applications: If your embedded system requires real-time performance or low-latency response times because `std` can introduce unpredictable delays and overhead that can affect real-time performance.
- Custom requirements: bare-metal allows more customization and fine-grained control over the behavior of an application, which can be useful in specialized or non-standard environments.
```

I think adding this section from @JurajSadel [blogpost](https://blog.espressif.com/rust-embedded-a-development-power-duo-db43dae21206) can be really useful

It's important to note that since `no_std`uses the Rust core library, a subset of the Rust standard library,  a `no_std` crate can compile in `std` environment but the oposite is not true. Therefore, when creating crates it's worth keeping in mind if it needs the standard library to function.

[embedded-rust-book]: https://docs.rust-embedded.org/
[embedded-rust-book-no-std]: https://docs.rust-embedded.org/book/intro/no-std.html
[rust-lib-core]: https://doc.rust-lang.org/core/index.html
[rust-lib-std]: https://doc.rust-lang.org/std/index.html

## Current support

The table below covers the current support for `no_std` at this moment for different Espressif products.

> **Notes**:
>
> - ✅ - the feature is implemented or supported
> - ⏳ - the feature is under development
> - ❌ - the feature is not supported
> - [ESP8266 HAL][esp-rs/esp8266-hal] is in maintenance mode and no further development will be done for this chip

> |         | [HAL][esp-rs/esp-hal] | [PAC][esp-rs/esp-pacs] | [Wi-Fi][esp-rs/esp-wifi] | [Bluetooth][esp-rs/esp-wifi] | [Heap][esp-rs/esp-alloc] | [Println][esp-rs/esp-println] | [Backtrace][esp-rs/esp-backtrace] | [Storage][esp-rs/esp-storage] |

Reviewer 2: I think this table is going to be tough to maintain and is going to fall out of sync with reality pretty fast. We will be adding at least 2 more rows this year, with more to come, and more columns will be added as well (eg. for IEEE802.15.4, ULP support, etc.).

If we do keep this then we should at least change `heap` to `alloc` in the header, IMO.

Reviewer: We had some conversations about how maintain this table. I dont have any hard opinion, but I think its a good thing to have, maybe we can remove some columns or merge them (only 1 column for esp-wif) and have a symbol that represents partial support that can be used in cases that, for example Wifi is supported but BLE is not.

| -------- | :---: | :---: | :---: | :---: | :---: | :---: | :---: | :---: |
|  ESP32   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |

> | ESP32-C2 |   ✅   |   ✅   |   ✅   |   ✅   |   ❌   |   ✅   |   ✅   |   ✅   |

Reviewer 2: I don't know why C2 and C6 are marked as not supported, but they are. `esp-alloc` doesn't even have features per-chip.

| ESP32-C3 |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |
| ESP32-C6 |   ✅   |   ✅   |   ✅   |   ⏳   |   ❌   |   ✅   |   ✅   |   ❌   |
| ESP32-S2 |   ✅   |   ✅   |   ✅   |   ❌   |   ✅   |   ✅   |   ✅   |   ✅   |
| ESP32-S3 |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |   ✅   |

> | ESP32-H2 |   ❌   |   ✅   |   ❌   |   ❌   |   ❌   |   ✅   |   ❌   |   ❌   |

Reviewer 2: Is there any point in including the H2 here when we don't really even have anything people can use?

Reviewer: Can be removed for the moment, or at least changed as in my PR to pending, but definitely not ❌

> | ESP8266  |   ✅   |   ✅   |   ❌   |   ❌   |   ❌   |   ✅   |   ❌   |   ❌   |

Reviewer 2: I think we should probably remove references to the ESP8266, none of the libraries for it are maintained.

Here is a list with the relevant crates and small summary of what they do for `no_std`:

| Symbol | Repository | Description |
| ------ | ---------- | ----------- |
| **HAL** | [esp-rs/esp-hal] | Hardware abstraction layer |
| **PAC** | [esp-rs/esp-pacs] | Peripheral access crates |
| **Wi-Fi** | [esp-rs/esp-wifi] | Wi-Fi support |

> | **Bluetooth** | [esp-rs/esp-wifi] | Bluetooth LE support |

Reviewer 2: Not sure what the "Symbol" column is for? Wi-Fi/Bluetooth use the same repo though, weird to link it twice, and ESP-NOW isn't mentioned at all (which is also supported by that repo).

Reviewer: We had a conversation around this in this comment: https://github.com/esp-rs/book/pull/75#discussion_r1153286243

Author: Delete the Symbol column
